lib/sycamore/sycamore/docset.py

```diff
@@ -975,28 +980,48 @@ def llm_filter(
             prompt: LLM prompt.
             field: Document field to filter based on.
             threshold: Cutoff that determines whether or not to keep document.
+            ignore_doc_structure: ignores document.element mapping.
+            similarity_query: query string to compute similarity against.
+            similarity_scorer: scorer to generate similarity if 'sort_elements_by_similarity' is True.
             **resource_args
 
         Returns:
             A filtered DocSet.
         """
-
-        def threshold_filter(doc: Document, threshold) -> bool:
-            try:
-                return_value = int(doc.properties[new_field]) >= threshold
-            except Exception:
-                # accounts for llm output errors
-                return_value = False
-
-            return return_value
-
-        docset = self.filter(lambda doc: doc.field_to_value(field) is not None and doc.field_to_value(field) != "None")
-
         entity_extractor = OpenAIEntityExtractor(
             entity_name=new_field, llm=llm, use_elements=False, prompt=prompt, field=field
         )
-        docset = docset.extract_entity(entity_extractor=entity_extractor)
-        docset = docset.filter(lambda doc: threshold_filter(doc, threshold), **resource_args)
+
+        def threshold_filter(doc: Document, threshold) -> bool:
+            if ignore_doc_structure:
+                if doc.field_to_value(field) is None:
+                    return False
+                doc = entity_extractor.extract_entity(doc)
+                return int(re.findall(r"\d+", doc.properties[new_field])[0]) >= threshold
+
+            # If we're looking at a document level custom property (i.e. not text_representation),
+            # only evaluate the Document and not elements.
+            if field != "text_representation" and doc.field_to_value(field) is not None:
+                doc = entity_extractor.extract_entity(doc)
+                return int(re.findall(r"\d+", doc.properties[new_field])[0]) >= threshold
+
+            if similarity_query or similarity_scorer:
+                assert similarity_scorer is not None, "Similarity sorting requires a scorer"
+                assert similarity_query is not None, "Similarity sorting requires a string query"
+                score_property_name = f"{field}_similarity_score"
+                doc = similarity_scorer.generate_similarity_scores(
+                    doc_batch=[doc], query=similarity_query, score_property_name=score_property_name
+                )[0]
+                doc.elements.sort(key=lambda e: e.properties.get(score_property_name, float("-inf")), reverse=True)
+            for element in doc.elements:
+                e_doc = Document(element.data)
```

Comment:
This is really the way to lift an element into a Document?

Reply:
Yeah 🙃